Specify broadcast=1 for Pow in onnx->caffe2 converter

kevinbchen opened the pull request:
Summary:
From onnx opset 7, Pow supports broadcast by default; unfortunately this
isn't the case for Pow in caffe2 yet. An onnx model that relies on Pow
broadcasting will not run when converted to caffe2 ("Dimension mismatch - did
you forget to set broadcast=1?").

This is a band-aid fix to simply add broadcast=1 to the caffe2 Pow operator
during the conversion. The proper fix is to have Pow support broadcast in
caffe2, and there is an old diff to do so - D8528654.

Differential Revision: D13830604

ezyang commented:
ASAN failure looks real:

```
Jan 26 03:01:45 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/functional:2267:14: runtime error: call to function std::_Function_handler<std::unique_ptr<caffe2::FunctionSchemaStorageBase, std::default_delete<caffe2::FunctionSchemaStorageBase> > (), std::unique_ptr<caffe2::FunctionSchemaStorageBase, std::default_delete<caffe2::FunctionSchemaStorageBase> > (*)()>::_M_invoke(std::_Any_data const&) through pointer to incorrect function type 'std::unique_ptr<caffe2::FunctionSchemaStorageBase, std::default_delete<caffe2::FunctionSchemaStorageBase> > (*)(const std::_Any_data &)'
Jan 26 03:01:45 (/opt/conda/lib/python3.6/site-packages/torch/lib/libcaffe2.so+0xb6eb8c0): note: std::_Function_handler<std::unique_ptr<caffe2::FunctionSchemaStorageBase, std::default_delete<caffe2::FunctionSchemaStorageBase> > (), std::unique_ptr<caffe2::FunctionSchemaStorageBase, std::default_delete<caffe2::FunctionSchemaStorageBase> > (*)()>::_M_invoke(std::_Any_data const&) defined here
Jan 26 03:01:45     #0 0x7f569db9bc1a in std::function<std::unique_ptr<caffe2::FunctionSchemaStorageBase, std::default_delete<caffe2::FunctionSchemaStorageBase> > ()>::operator()() const (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch.so.1+0x23efc1a)
Jan 26 03:01:45     #1 0x7f569db9b1dc in c10::Registry<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unique_ptr<caffe2::FunctionSchemaStorageBase, std::default_delete<caffe2::FunctionSchemaStorageBase> >>::Create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch.so.1+0x23ef1dc)
Jan 26 03:01:45     #2 0x7f569db992aa in torch::jit::createOperatorFromCaffe2(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch.so.1+0x23ed2aa)
Jan 26 03:01:45     #3 0x7f569dba2baa in torch::jit::RegisterOperators::Caffe2Operator(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch.so.1+0x23f6baa)
Jan 26 03:01:45     #4 0x7f569cc05ec1 in _GLOBAL__sub_I_register_caffe2_ops.cpp (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch.so.1+0x1459ec1)
Jan 26 03:01:45     #5 0x7f56cc7416b9 in call_init /build/glibc-Cl5G7W/glibc-2.23/elf/dl-init.c:72
Jan 26 03:01:45     #6 0x7f56cc7417ca in call_init /build/glibc-Cl5G7W/glibc-2.23/elf/dl-init.c:30
Jan 26 03:01:45     #7 0x7f56cc7417ca in _dl_init /build/glibc-Cl5G7W/glibc-2.23/elf/dl-init.c:120
Jan 26 03:01:45     #8 0x7f56cc7468e1 in dl_open_worker /build/glibc-Cl5G7W/glibc-2.23/elf/dl-open.c:575
Jan 26 03:01:45     #9 0x7f56cc741563 in _dl_catch_error /build/glibc-Cl5G7W/glibc-2.23/elf/dl-error.c:187
Ja
```

ezyang commented:
I'm not sure how this PR could have possibly triggered this particular failure though

Commit:
Specify broadcast=1 for Pow in onnx->caffe2 converter (#16396)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16396

From onnx opset 7, Pow supports broadcast by default; unfortunately this
isn't the case for Pow in caffe2 yet. An onnx model that relies on Pow
broadcasting will not run when converted to caffe2 ("Dimension mismatch - did
you forget to set broadcast=1?").

This is a band-aid fix to simply add broadcast=1 to the caffe2 Pow operator
during the conversion. The proper fix is to have Pow support broadcast in
caffe2, and there is an old diff to do so - D8528654.

Differential Revision: D13830604

fbshipit-source-id: f3338c850fb3c0f12847d01437bb33beb285160a

garymm commented:
Should we close this?
